Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,362 @@
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 1);
+    EXPECT_LE(now(), far_away);
+  });
+
+  std::this_thread::sleep_for(chr::milliseconds(5));
+  schedule.Push(2, far_away);
+  std::this_thread::sleep_for(chr::milliseconds(1));
+  schedule.Push(1, start_time + chr::milliseconds(100));
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(ScheduleTest, PopBlockingNoticesRemovals) {
+  const auto future = std::async(std::launch::async, [&] {
+    schedule.Push(1, start_time + chr::milliseconds(50));
+    schedule.Push(2, start_time + chr::milliseconds(100));
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 2);
+  });
+
+  while (schedule.empty()) {
+    std::this_thread::sleep_for(chr::milliseconds(1));
+  }
+  const bool removed =
+      schedule.RemoveIf(nullptr, [](const int v) { return v == 1; });
+  EXPECT_TRUE(removed);
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(ScheduleTest, PopBlockingIsNotAffectedByIrrelevantRemovals) {
+  const auto future = std::async(std::launch::async, [&] {
+    schedule.Push(1, start_time + chr::milliseconds(50));
+    schedule.Push(2, start_time + chr::seconds(10));
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 1);
+  });
+
+  while (schedule.empty()) {
+    std::this_thread::sleep_for(chr::milliseconds(1));
+  }
+  const bool removed =
+      schedule.RemoveIf(nullptr, [](const int v) { return v == 2; });
+  EXPECT_TRUE(removed);
+  ABORT_ON_TIMEOUT(future);
+}
+
+// AsyncQueue tests
+
+namespace {
+
+class AsyncQueueTest : public ::testing::Test {
+ public:
+  AsyncQueueTest() : signal_finished{[] {}} {
+  }
+
+  // Googletest doesn't contain built-in functionality to block until an async
+  // operation completes, and there is no timeout by default. Work around both
+  // by resolving a packaged_task in the async operation and blocking on the
+  // associated future (with timeout).
+  bool WaitForTestToFinish(const chr::milliseconds timeout = kTimeout) {
+    return WaitForFuture(signal_finished.get_future(), timeout);
+  }
+
+  // Used in async tests to notify the main thread that an opration has
+  // finished.
+  std::packaged_task<void()> signal_finished;
+  AsyncQueue queue;
+};
+
+}  // namespace
+
+TEST_F(AsyncQueueTest, Enqueue) {
+  queue.Enqueue([&] { signal_finished(); });
+  EXPECT_TRUE(WaitForTestToFinish());
+}
+
+TEST_F(AsyncQueueTest, DestructorDoesNotBlockIfThereArePendingTasks) {
+  const auto future = std::async(std::launch::async, [&] {
+    AsyncQueue another_queue;
+    another_queue.EnqueueAfterDelay(chr::minutes(5), [] {});
+    another_queue.EnqueueAfterDelay(chr::minutes(10), [] {});
+    // Destructor shouldn't block waiting for the 5/10-minute-away operations.
+  });
+
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(AsyncQueueTest, CanScheduleOperationsInTheFuture) {
+  std::string steps;
+
+  queue.Enqueue([&steps] { steps += '1'; });
+  queue.EnqueueAfterDelay(AsyncQueue::Milliseconds(5), [&] {
```

Comment:
@rsgowman The timing issue you pointed out applies here as well. I'd like a similar solution for both versions of `AsyncQueue` (though for C++-based one, there is considerably more control over clock that we have).

Reply:
Yeah, as pointed out above, you could inject a fake clock into this solution. I don't think you can easily do that for libdispatch though.

Using absolute times combined with the ability to pause execution on the queue might be viable.